### ISSUE TITLE
Should the use of both show and hide on an import be a syntax error?

### ISSUE THREAD

**Issue by alexeyinkin** [dart-lang/language#2824]
In this example `hide` adds no meaning because if an identifier is not in the list of `show` it is not visible anyway:

```dart
import 'package:collection/collection.dart' show ListEquality hide MapEquality;

void main() {
  ListEquality();
  //MapEquality(); // undefined
}
```

Moreover an identifier can be in both `show` and `hide` resulting in it being hidden:

```dart
import 'package:collection/collection.dart' show ListEquality hide ListEquality; // ignore: unused_import

void main() {
  //ListEquality(); // undefined
}
```

Both of these examples build and produce no linter issues.

Is there a valid case for using both `show` and `hide` on a single import? If yes, then there should be a lint that flags the meaningless cases like above. If no, should this be a syntax error? With Dart 3 allowed to introduce breaking changes, should this opportunity be used to add this syntax change?

**Comment by incendial** [comment]
I'd also add that it's currently possible to have multiple show / hide with the same identifier, like:

```dart
import 'package:collection/collection.dart' show ListEquality show ListEquality show ListEquality show ListEquality;

void main() {
  ListEquality();
}
```
and it produces no errors, but looks strange.

**Comment by eernstg** [comment]
Agreed! - the semantics of multiple show and hide clauses on imports and exports is convoluted and even counter-intuitive.

I argued in https://github.com/dart-lang/sdk/issues/32878 that `show ... show ...` should provide access to the union of the lists of names in those `show` clauses, rather than the intersection (which is actually the outcome according to the specification). The behavior of `show ... hide ...` is similarly counter-intuitive.

However, these rules have been around for more than a decade, and there wasn't much appetite in 2018 (or indeed at any time) to change anything in this area. This was probably so because nobody felt a strong urge to write powerful and complex `show ... hide ...` clauses anyway, and there's always a cost associated with the introduction of breaking changes.

I've created a lint proposal, https://github.com/dart-lang/linter/issues/4052.

**Comment by lrhn** [comment]
I'd happily support only allowing one `hide` or `show` on each import. There is never need for anything more.

If we ever introduce a "rename" combinator, we might need to reconsider, but until then one is always enough. We're not helping anyone by allowing more.

(That `show` works like intersection is a consequence of it really meaning "hide-everything-but". Showing everything is the default, so we only need hide-operators. One of them is called `show`.)

The alternative is to do nothing, which has worked so far. I haven't seen anyone having any real problems, which also makes me think adding a lint is overkill. If we do anything, it can just be a plain analyzer warning, there is never a reason to turn it off.

**Comment by alexeyinkin** [comment]
If it is made a compile-time error, there will be one less complicated trailing comma case: https://github.com/dart-lang/language/issues/1402#issuecomment-913494400

**Comment by srawlins** [comment]
Ooh this could be a candidate for Dart 4.0.

**Comment by munificent** [comment]
This has always been a needlessly over-engineered corner of the language. My understanding is that the mental model is *supposed* to be:

* An import produces a set of imported declarations.
* Then each combinator in turn acts upon that set and produces a new set. A `show` combinator takes the incoming set of declarations and strips it down to just the shown names. A `hide` combinator subtracts the hidden names and leaves the rest.
* The combinators are processed as an ordered pipeline of set operators.

This is why they are called "combinators". The idea is they can be freely combined to manipulate name sets however you want.

But... there are literally no *useful* combinations of these combinators:

* `show` followed by another `show` can always be expressed more tersely as a single `show`. Instead of `show A, B, C, D show B, D` just write `show B, D`.
* `show` followed by `hide` can always be expressed more tersely by deleting the hidden names from the `show` and eliminating the `hide`. Instead of `show A, B, C, D hide A, C`, just write `show B, D`.
* `hide` followed by another `hide` can always be combined into a single `hide`. Instead of `hide A hide B` just write `hide A, B`.
* `hide` followed by `show` can again always be expressed more tersely by deleting the hidden names from the `show` and eliminating `hide`. Instead of `hide A, C show A, B, C, D`, just write `show B, D`.

I honestly don't understand why the feature was designed this way. Maybe the initial designers thought we would add other combinators for more complex namespace manipulation later?

**Comment by lrhn** [comment]
> Maybe the initial designers thought we would add other combinators for more complex namespace manipulation later?

They did. Renaming was one option that was always considered. 

But I'm still not sure you'd need both `show` and `hide`, even if you had rename.

**Comment by munificent** [comment]
We discussed this at the language meeting and we are all in favor of fixing this and accepting @srawlins's proposal.

This would be a language-versioned change. Multiple combinators would still be allowed in language versions prior to the version this ships in. Since it's language versioned, we can ship it in a minor version or in Dart 4.0.

We could make this an error in *all* language versions and ship that in Dart 4.0. But the feeling on the language team is that it would be better to language-version the change. It feels a little weird to language version this since it isn't really a new "feature". It's simply removing a (useless) capability. But @eernstg pointed out that *simplicity* is a feature, so this is still a language improvement. Given that, I think it makes sense to language version it.

There is already a warning (not a lint) which defaults to on if you try to combine combinators. The proposal then is that that remains a warning in older language versions and then becomes an error in the language version we ship this change in.

We haven't decided if implementation work should begin on this or what we release we intend to make it in. That's above my pay grade and @leafpetersen was out this week. :)

**Comment by munificent** [comment]
We should probably wait until @leafpetersen is back first, but assuming he is OK with this, @srawlins, you can move the proposal to `accepted/`.

**Comment by leafpetersen** [comment]
> We could make this an error in _all_ language versions and ship that in Dart 4.0

Not unless we drop support for all previous language versions in Dart 4 right?  Which I think is an unlikely path for us. [Edit: Rather, we would have to block all packages with an upper bound of <4 from resolving - we could still accept previous language versions, but every package would explicitly have to test and bump their upper bound]

> We haven't decided if implementation work should begin on this or what we release we intend to make it in.

I'd suggest that we target 4.  In either case, fine with implementation work beginning whenever it's appropriate for the affected teams.

> We should probably wait until [@leafpetersen[<img alt="" width="15" height="15" src="chrome-extension://dlebflppeeemcdpidccbiblndppbmjmh/ospo-chrome-ext-logo.png">](https://teams.googleplex.com/leafp@google.com)](https://github.com/leafpetersen) is back first, but assuming he is OK with this

LGTM, thanks!

**Comment by munificent** [comment]
> Not unless we drop support for all previous language versions in Dart 4 right?

What I mean is that we would literally change the behavior of all language versions. It would just spontaneously become an error, even in <4.0 code. Not something I'd want to do, but it came up in the meeting so I wanted to mention the trade-offs.

**Comment by leafpetersen** [comment]
> What I mean is that we would literally change the behavior of all language versions. It would just spontaneously become an error, even in <4.0 code.

Yes, I understood that.  But that's incompatible with continuing to support earlier language versions, unless we decide to just break packages on pub, which I don't know why we would do for this specifically when we're trying to avoid it elsewhere?

**Comment by munificent** [comment]
>  I don't know why we would do for this specifically when we're trying to avoid it elsewhere?

Agreed! It came up because I think someone not in the meeting at had previously mentioned that it seemed weird to make a language versioned change that wasn't really *adding* a feature. So we discussed the idea briefly of just making it a hard breaking change in all language versions, but I think we all agree that wouldn't be a good idea.
